src/osp/tasks/execute_simple.cpp

```diff
@@ -0,0 +1,193 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "execute_simple.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+#include <algorithm>
+#include <cassert>
+
+namespace osp
+{
+
+void task_enqueue(TaskTags const& tags, ExecutionContext &rExec, BitSpanConst_t const query)
+{
+    std::size_t const tagIntSize = tags.tag_ints_per_task();
+    assert(query.size() == tagIntSize);
+
+    auto const taskTagInts = Corrade::Containers::arrayView(tags.m_taskTags);
+
+    for (uint32_t const currTask : tags.m_tasks.bitview().zeros())
+    {
+        unsigned int &rQueuedCount = rExec.m_taskQueuedCounts[currTask];
+
+        if (rQueuedCount == 0)
+        {
+            // Task not yet queued, check if any tags match the query
+            std::size_t const offset = currTask * tagIntSize;
+            auto const currTaskTagInts = taskTagInts.slice(offset, offset + tagIntSize);
+
+            bool anyTagMatches = false;
+            auto taskTagIntIt = std::begin(currTaskTagInts);
+
+            for (uint64_t const queryInt : query)
+            {
+                if ((queryInt & *taskTagIntIt) != 0)
+                {
+                    anyTagMatches = true;
+                    break;
+                }
+                std::advance(taskTagIntIt, 1);
+            }
+
+            if (anyTagMatches)
+            {
+                rQueuedCount = 1;
+                auto const view = lgrn::bit_view(currTaskTagInts);
+                for (uint32_t const tag : view.ones())
+                {
+                    rExec.m_tagIncompleteCounts[tag] ++;
+                }
+            }
+        }
+    }
+}
+
+static bool compare_tags(BitSpanConst_t const mask, BitSpanConst_t const taskTags) noexcept
+{
+    auto taskTagIntIt = std::begin(taskTags);
+    for (uint64_t const maskInt : mask)
+    {
+        uint64_t const taskTagInt = *taskTagIntIt;
+
+        // No match if a 1 bit in taskTagInt corresponds with a 0 in maskInt
+        if ((maskInt & taskTagInt) != taskTagInt)
+        {
+            return false;
+        }
+        std::advance(taskTagIntIt, 1);
+    }
+    return true;
+}
+
+void task_list_available(TaskTags const& tags, ExecutionContext const& exec, BitSpan_t tasksOut)
+{
+    assert(tasksOut.size() == tags.m_tasks.vec().size());
+
+    // Bitmask makes it easy to compare the tags of a task
+    // 1 = allowed (default), 0 = not allowed
+    // All tags of a task must be allowed for the entire task to run.
+    // aka: All ones in a task's bits must corrolate to a one in the mask
+    std::vector<uint64_t> mask(tags.m_tags.vec().size(), ~uint64_t(0));
```

Comment:
std::numeric_limits<uint64_t>::max()

~uint64_t(0) is not the same on all platforms.

Reply:
This int is intended to have all the bits set to 1, and not a max value.

I'm aware there's undefined behaviour associated with signed integers, but I'm unsure of how `~uint64_t(0) ` can be problematic?

Reply:
reading comprehension fail on my part. I see you had a comment explaining that. Never mind!